cupyx/scipy/signal/_peak_finding.py

```diff
@@ -1492,3 +1497,659 @@ def argrelextrema(data, comparator, axis=0, order=1, mode="clip"):
+    [[19.200000000000003 14.6 --]
+     [40.0 37.5 --]
+     [42.800000000000004 40.05 --]]
+
+    """
+    def _quantiles1D(data, m, p):
+        x = np.sort(data.compressed())
+        n = len(x)
+        if n == 0:
+            return ma.array(np.empty(len(p), dtype=float), mask=True)
+        elif n == 1:
+            return ma.array(np.resize(x, p.shape), mask=nomask)
+        aleph = (n*p + m)
+        k = np.floor(aleph.clip(1, n-1)).astype(int)
+        gamma = (aleph-k).clip(0, 1)
+        return (1.-gamma)*x[(k-1).tolist()] + gamma*x[k.tolist()]
+
+    data = ma.array(a, copy=False)
+    if data.ndim > 2:
+        raise TypeError("Array should be 2D at most !")
+
+    if limit:
+        condition = (limit[0] < data) & (data < limit[1])
+        data[~condition.filled(True)] = masked
+
+    p = np.array(prob, copy=False, ndmin=1)
+    m = alphap + p*(1.-alphap-betap)
+    # Computes quantiles along axis (or globally)
+    if (axis is None):
+        return _quantiles1D(data, m, p)
+
+    return ma.apply_along_axis(_quantiles1D, axis, data, m, p)
+
+
+def _scoreatpercentile(data, per, limit=(), alphap=.4, betap=.4):
+    """Calculate the score at the given 'per' percentile of the
+    sequence a.  For example, the score at per=50 is the median.
+
+    This function is a shortcut to mquantile
+
+    """
+    if (per < 0) or (per > 100.):
+        raise ValueError("The percentile should be between 0. and 100. !"
+                         " (got %s)" % per)
+
+    return _mquantiles(data, prob=[per/100.], alphap=alphap, betap=betap,
+                       limit=limit, axis=0).squeeze()
+
+
+_NOISE_MODULE_SRC = r"""
+#include <cstddef>  // for std::size_t
+#include <type_traits>
+
+#include <thrust/sort.h>
+#include <thrust/functional.h>
+#include <thrust/execution_policy.h>
+
+__device__ inline float clip(float value, float lb, float ub) {
+    return value < lb ? lb : ((value > ub) ? ub : value);
+}
+
+template<typename T>
+__global__ cwt_noise(
+        const int n_points,
+        const int window_size,
+        const float prob,
+        const float m,
+        const T* __restrict__ cwt_row,
+        T* out) {
+
+    int idx = blockDim.x * blockIdx.x + threadIdx.x;
+
+    if(idx >= n_points) {
+        return;
+    }
+
+    const int win_half_size = window_size / 2;
+    const int odd = window_size % 2;
+
+    const int left_pos = min(idx - win_half_size, 0);
+    const int right_pos = min(idx + win_half_size + odd, n);
+    const int n = right_pos - left_pos;
+
+    thrust::sort(thrust::device, cwt_row + left_pos, cwt_row + right_pos);
+
+    float aleph = n * prob + m;
+    int k = floor(clip(aleph, 1, n - 1));
+    float gamma = clip(aleph - k, 0, 1);
+
+    const T* x = cwt_row + left_pos;
+    out[idx] = (1.0 - gamma) * x[k - 1] + gamma * x[k];
+}
+"""
+
+_noise_module = cupy.RawModule(
+    code=_NOISE_MODULE_SRC, options=('-std=c++11',),
+    jitify=True,
+    name_expressions=['cwt_noise<double>', 'cwt_noise<float>'])
```

Comment:
@takagi, do you know which arguments should be passed into `RawModule` in order to support `thrust`?

Reply:
I tried this branch on my local, and have (perhaps cccl related?) errors both with and without jitify. I'm looking into this further.

Reply:
It is supposed that jitify cannot find Thrust headers.
```
thrust/functional.h(26): warning: functional: [jitify] File not found
thrust/system/cuda/detail/util.h(31): warning: exception: [jitify] File not found
```

The following could be compiled successfully, but the error happens if `option` is omitted. (Thanks, @.kmaehashi)
```
import cupy
m = cupy.RawKernel('''
#include <thrust/functional.h>

extern "C" __global__ void test(){}
''', 'test', jitify=True, options=('-I/path/to/cupy/cupy/_core/include/cupy/_cccl/thrust',))
m.compile()
```

@leofang Do you have any thought?

Reply:
@takagi sorry I missed the ping... Did any actual error happen? The "File not found" warning is harmless, and is suppressed in #8235. Setting `options` as you did above should never be needed; if it is, it's a bug.

Reply:
The case may have changed. Please ignore setting `options`. Now I've got the following log with making jitify logging alive. My local is on top of `andfoy:add_find_peaks_cwt`, merged current `main` and is ignoring wavelet's deprecation warnings. 
```
-------------------------------------------------------- Captured stdout call --------------------------------------------------------
thrust/functional.h(26): warning: functional: [jitify] File not found
thrust/system/cuda/detail/util.h(31): warning: exception: [jitify] File not found
---------------------------------------------------
--- JIT compile log for /tmp/tmpuwt1zfy5/00fb6c5cabaaec76a86f14cf490998b19ef611c9.cubin.cu ---
---------------------------------------------------
thrust/detail/config/cpp_dialect.h(132): warning #161-D: unrecognized #pragma
       THRUST_COMPILER_DEPRECATION_SOFT(C++14, C++11);
       ^

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

cub/util_cpp_dialect.cuh(143): warning #161-D: unrecognized #pragma
       CUB_COMPILER_DEPRECATION_SOFT(C++14, C++11);
       ^

thrust/detail/alignment.h(154): warning #1866-D: attribute does not apply to any entity
          struct type {} __attribute__((aligned(Align)));
                                        ^

thrust/detail/static_assert.h(49): warning #47-D: incompatible redefinition of macro "THRUST_STATIC_ASSERT" (declared at line 9 of jitify_preinclude.h)
  #    define THRUST_STATIC_ASSERT(B)        static_assert(B, "static assertion failed")
              ^

thrust/system/error_code.h(185): error: namespace "std" has no member "string"
      virtual std::string message(int ev) const = 0;
                   ^

thrust/system/error_code.h(297): error: namespace "std" has no member "string"
      inline std::string message(void) const;
                  ^

thrust/system/error_code.h(419): error: namespace "std" has no member "string"
      inline std::string message(void) const;
                  ^

thrust/system/detail/error_category.inl(98): error: namespace "std" has no member "string"
      inline virtual std::string message(int ev) const
                          ^

thrust/system/detail/error_category.inl(100): error: namespace "std" has no member "string"
        static const std::string unknown_err("Unknown error");
                          ^

thrust/system/detail/error_category.inl(107): error: namespace "std" has no member "string"
        return c_str ? std::string(c_str) : unknown_err;
                            ^

thrust/system/detail/error_category.inl(123): error: namespace "std" has no member "string"
      inline virtual std::string message(int ev) const
                          ^

thrust/system/detail/error_code.inl(111): error: namespace "std" has no member "string"
  std::string error_code
       ^

thrust/system/detail/error_condition.inl(105): error: namespace "std" has no member "string"
  std::string error_condition
       ^

stdexcept(6): error: namespace "std" has no member "string"
  explicit runtime_error( const std::string& what_arg );explicit runtime_error( const char* what_arg );virtual const char* what() const;
                                     ^

thrust/system/cuda/detail/guarded_driver_types.h(30): warning #161-D: unrecognized #pragma
  #    pragma push_macro("__host__")
              ^

thrust/system/cuda/detail/guarded_driver_types.h(35): warning #161-D: unrecognized #pragma
  #    pragma push_macro("__device__")
              ^

thrust/system/cuda/detail/guarded_driver_types.h(56): warning #161-D: unrecognized #pragma
  #    pragma pop_macro("__host__")
              ^

thrust/system/cuda/detail/guarded_driver_types.h(60): warning #161-D: unrecognized #pragma
  #    pragma pop_macro("__device__")
              ^

thrust/system/cuda/detail/error.inl(69): error: no operator "+" matches these operands
            operand types are: std::string + const char [3]
             + ": " + (c_str ? c_str : unknown_str);
             ^

cub/detail/device_synchronize.cuh(39): error: this declaration has no storage class or type specifier
  CUB_RUNTIME_FUNCTION inline cudaError_t device_synchronize()
  ^

cub/detail/device_synchronize.cuh(39): error: expected a ";"
  CUB_RUNTIME_FUNCTION inline cudaError_t device_synchronize()
                       ^

cub/detail/device_synchronize.cuh(61): warning #12-D: parsing restarts here after previous syntax error
    NV_IF_TARGET(NV_IS_HOST,
    ^

cub/detail/device_synchronize.cuh(67): error: expected a declaration
    return result;
    ^

cub/detail/device_synchronize.cuh(72): error: expected a declaration
  CUB_NAMESPACE_END
  ^

stdint.h(29): warning #12-D: parsing restarts here after previous syntax error
  typedef unsigned long long uint_fast64_t;
                                          ^

stdint.h(36): error: invalid redeclaration of type name "intptr_t" (declared at line 43 of __cuda/cstdint_prelude.h)
   typedef signed long      intptr_t; //optional
                            ^

stdint.h(69): error: invalid redeclaration of type name "uintptr_t" (declared at line 44 of __cuda/cstdint_prelude.h)
    typedef unsigned long      uintptr_t; //optional
                               ^

stdint.h(72): error: expected a declaration
  } // namespace __jitify_stdint_ns
  ^

cuda_fp16.h(177): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __double2half(const double a);
  ^

cuda_fp16.h(177): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __double2half(const double a);
  ^

cuda_fp16.h(192): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half(const float a);
  ^

cuda_fp16.h(192): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half(const float a);
  ^

cuda_fp16.h(192): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half(const float a);
  ^

cuda_fp16.h(207): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rn(const float a);
  ^

cuda_fp16.h(207): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rn(const float a);
  ^

cuda_fp16.h(207): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rn(const float a);
  ^

cuda_fp16.h(222): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rz(const float a);
  ^

cuda_fp16.h(222): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rz(const float a);
  ^

cuda_fp16.h(222): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rz(const float a);
  ^

cuda_fp16.h(238): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rd(const float a);
  ^

cuda_fp16.h(238): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rd(const float a);
  ^

cuda_fp16.h(238): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_rd(const float a);
  ^

cuda_fp16.h(254): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_ru(const float a);
  ^

cuda_fp16.h(254): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_ru(const float a);
  ^

cuda_fp16.h(254): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __float2half_ru(const float a);
  ^

cuda_fp16.h(269): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __half2float(const __half a);
  ^

cuda_fp16.h(269): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __half2float(const __half a);
  ^

cuda_fp16.h(269): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ float __half2float(const __half a);
  ^

cuda_fp16.h(287): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float2half2_rn(const float a);
  ^

cuda_fp16.h(287): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float2half2_rn(const float a);
  ^

cuda_fp16.h(287): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float2half2_rn(const float a);
  ^

cuda_fp16.h(308): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __floats2half2_rn(const float a, const float b);
  ^

cuda_fp16.h(308): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __floats2half2_rn(const float a, const float b);
  ^

cuda_fp16.h(308): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __floats2half2_rn(const float a, const float b);
  ^

cuda_fp16.h(324): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __low2float(const __half2 a);
  ^

cuda_fp16.h(324): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __low2float(const __half2 a);
  ^

cuda_fp16.h(324): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ float __low2float(const __half2 a);
  ^

cuda_fp16.h(340): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __high2float(const __half2 a);
  ^

cuda_fp16.h(340): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ float __high2float(const __half2 a);
  ^

cuda_fp16.h(340): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ float __high2float(const __half2 a);
  ^

cuda_fp16.h(356): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ short int __half2short_rz(const __half h);
  ^

cuda_fp16.h(356): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ short int __half2short_rz(const __half h);
  ^

cuda_fp16.h(356): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ short int __half2short_rz(const __half h);
  ^

cuda_fp16.h(373): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned short int __half2ushort_rz(const __half h);
  ^

cuda_fp16.h(373): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned short int __half2ushort_rz(const __half h);
  ^

cuda_fp16.h(373): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned short int __half2ushort_rz(const __half h);
  ^

cuda_fp16.h(389): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ int __half2int_rz(const __half h);
  ^

cuda_fp16.h(389): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ int __half2int_rz(const __half h);
  ^

cuda_fp16.h(389): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ int __half2int_rz(const __half h);
  ^

cuda_fp16.h(405): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned int __half2uint_rz(const __half h);
  ^

cuda_fp16.h(405): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned int __half2uint_rz(const __half h);
  ^

cuda_fp16.h(405): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned int __half2uint_rz(const __half h);
  ^

cuda_fp16.h(421): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ long long int __half2ll_rz(const __half h);
  ^

cuda_fp16.h(421): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ long long int __half2ll_rz(const __half h);
  ^

cuda_fp16.h(421): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ long long int __half2ll_rz(const __half h);
  ^

cuda_fp16.h(438): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned long long int __half2ull_rz(const __half h);
  ^

cuda_fp16.h(438): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned long long int __half2ull_rz(const __half h);
  ^

cuda_fp16.h(438): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ unsigned long long int __half2ull_rz(const __half h);
  ^

cuda_fp16.h(460): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float22half2_rn(const float2 a);
  ^

cuda_fp16.h(460): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float22half2_rn(const float2 a);
  ^

cuda_fp16.h(460): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half2 __float22half2_rn(const float2 a);
  ^

cuda_fp16.h(476): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ float2 __half22float2(const __half2 a);
  ^

cuda_fp16.h(476): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ float2 __half22float2(const __half2 a);
  ^

cuda_fp16.h(476): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ float2 __half22float2(const __half2 a);
  ^

cuda_fp16.h(492): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ int __half2int_rn(const __half h);
  ^

cuda_fp16.h(492): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ int __half2int_rn(const __half h);
  ^

cuda_fp16.h(492): error: expected a ";"
  __CUDA_FP16_DECL__ int __half2int_rn(const __half h);
  ^

cuda_fp16.h(508): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ int __half2int_rd(const __half h);
  ^

cuda_fp16.h(508): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ int __half2int_rd(const __half h);
  ^

cuda_fp16.h(508): error: expected a ";"
  __CUDA_FP16_DECL__ int __half2int_rd(const __half h);
  ^

cuda_fp16.h(524): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ int __half2int_ru(const __half h);
  ^

cuda_fp16.h(524): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ int __half2int_ru(const __half h);
  ^

cuda_fp16.h(524): error: expected a ";"
  __CUDA_FP16_DECL__ int __half2int_ru(const __half h);
  ^

cuda_fp16.h(541): error: identifier "__host__" is undefined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __int2half_rn(const int i);
  ^

cuda_fp16.h(541): error: variable "__device__" has already been defined
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __int2half_rn(const int i);
  ^

cuda_fp16.h(541): error: expected a ";"
  __CUDA_HOSTDEVICE_FP16_DECL__ __half __int2half_rn(const int i);
  ^

cuda_fp16.h(557): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ __half __int2half_rz(const int i);
  ^

cuda_fp16.h(557): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ __half __int2half_rz(const int i);
  ^

cuda_fp16.h(557): error: expected a ";"
  __CUDA_FP16_DECL__ __half __int2half_rz(const int i);
  ^

cuda_fp16.h(573): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ __half __int2half_rd(const int i);
  ^

cuda_fp16.h(573): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ __half __int2half_rd(const int i);
  ^

cuda_fp16.h(573): error: expected a ";"
  __CUDA_FP16_DECL__ __half __int2half_rd(const int i);
  ^

cuda_fp16.h(589): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ __half __int2half_ru(const int i);
  ^

cuda_fp16.h(589): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ __half __int2half_ru(const int i);
  ^

cuda_fp16.h(589): error: expected a ";"
  __CUDA_FP16_DECL__ __half __int2half_ru(const int i);
  ^

cuda_fp16.h(607): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ short int __half2short_rn(const __half h);
  ^

cuda_fp16.h(607): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ short int __half2short_rn(const __half h);
  ^

cuda_fp16.h(607): error: expected a ";"
  __CUDA_FP16_DECL__ short int __half2short_rn(const __half h);
  ^

cuda_fp16.h(623): error: explicit type is missing ("int" assumed)
  __CUDA_FP16_DECL__ short int __half2short_rd(const __half h);
  ^

cuda_fp16.h(623): error: variable "__device__" has already been defined
  __CUDA_FP16_DECL__ short int __half2short_rd(const __half h);
  ^

Error limit reached.
100 errors detected in the compilation of "/tmp/tmpuwt1zfy5/00fb6c5cabaaec76a86f14cf490998b19ef611c9.cubin.cu".
Compilation terminated.

---------------------------------------------------
```

Reply:
I'll try to check tomorrow. Is it reproducible with https://github.com/cupy/cupy/pull/7875#discussion_r1448356844?

Reply:
Sorry I dropped the ball... I can reproduce, checking now...

Reply:
I think these headers can only be compiled by nvcc, see https://github.com/cupy/cupy/blob/150c903890bc14ec6c2432813374d9fde1b7528c/cupyx/jit/thrust.py#L10-L14

Reply:
@andfoy, I may not understand the latest change of this PR, but do we need these headers: "thrust/execution_policy.h" and "thrust/functional.h"?

Reply:
@takagi, I'm sorry for the confusion, right now this PR needs a lot of cleaning. I'm working on an alternative version of the filter kernel to not use an external dependency sort that has not been pushed yet.

Reply:
Okay. Thanks for letting me know that!